##### RNASEQ ANALYSIS OF YBR MUTANTS
#### DONE WITH AARON'S LIST OF ORFS
### DATA GENERATED BY APRIL ON 06/13/22

##### GATHER AND SAVE DATA FROM APRIL's DIRECTORY
# coldata1 <- read.csv('~/RNASeq/YBR/fastq_files/YBR2/samples.csv', stringsAsFactors = FALSE, row.names = NULL)
# colnames(coldata1) <- c('filename','name')
# # coldata1 <- coldata1[str_detect(coldata1$filename, 'S'),]
# 
# coldata2 <- read.csv('~/RNASeq/YBR/fastq_files/YBR1/samples.csv', stringsAsFactors = FALSE, row.names = NULL)
# coldata2$mutant[coldata2$mutant == 'CRISPY-ATG'] <- "ATG"
# colnames(coldata2) <- c('filename','name')
# 
# coldata <- rbind(coldata1, coldata2)
# coldata$filename <- str_replace_all(coldata$filename, '-','.')
# 
# alignmentinfo <- readRDS('/home/aar75/rna_seq/Salmon_4_6_22/20220613_labSamplesAlignmentInfo.RDS')
# orfs <- readRDS('/home/aar75/rna_seq/Salmon_4_6_22/translatome.rds')
# rawcounts <- readRDS('/home/aar75/rna_seq/Salmon_4_6_22/20220613_labSamplesRawCounts.RDS')
# save(coldata, alignmentinfo, orfs, rawcounts, file = 'input/rnaseek.RData')

##### INITIALIZE
source('~/R/Projects/ybr/scripts/initialize.R')
load(file = '~/R/Projects/ybr/input/rnaseek.RData')

## COIP AND COEX DATA
ybr.coip <- read.csv('~/R/Projects/ybr/output/ybr_pd_scores.csv')
colnames(ybr.coip)
ybr.coip <- ybr.coip[,c(-1,-7)]
ybr.coip$coip_rank <- as.numeric(rownames(ybr.coip))

ybr.coex <- readRDS('/home/aar75/coexpression/20220607_ybr_coexpression.RDS')

# temp <- bitr(ybr.coex$gene, fromType = "ENSEMBL",
#              toType = c("ENTREZID","GENENAME"),
#              OrgDb = org.Sc.sgd.db)
# ybr.coex <- merge(ybr.coex, temp, by.x = 'gene', by.y = 'ENSEMBL', all.x = T)
ybr.coex <- merge(ybr.coex, ybr.coip[,-1],
                  by.x = 'gene',
                  by.y = 'ORF',
                  all = T)
ybr.coex <- ybr.coex[order(ybr.coex$rho, decreasing = T),]
rownames(ybr.coex) <- NULL
ybr.coex %>%
  filter(transcript %in% c('chr2_614215', 'chr2_614702', 'chr2_614936'))

## RNASEQ DATA
coldata$attempt[str_detect(coldata$filename, 'ARC')] <- 'One'
coldata$attempt[is.na(coldata$attempt)] <- 'Two'

coldata$owner[str_detect(coldata$filename, 'AW')] <- 'AW' #YEL021W
coldata$owner[is.na(coldata$owner)] <- 'SP'

temp <- merge(data.frame(filename = colnames(rawcounts)), coldata, by = 'filename')
temp <- temp %>% mutate(rep = rep(c(1,2,3),14))
colnames(rawcounts) <- c('transcript', paste0(temp$name, '_', temp$owner, '_', temp$attempt, '_', temp$rep))

merge(orfs[,c('orf_id','transcript','gene_systematic_name')] %>%
  filter(orf_id %in% c(143103, 143104, 143113, 143114)),
  rawcounts, by = 'transcript')

##### RAW CORRELATION
# raw.cor <- cor(rawcounts[,-1])
# data.frame(raw.cor)
# col1 <- colorRampPalette(c("#7F0000", "red", "#FF7F00", "yellow", "white",
#                            "cyan", "#007FFF", "blue","#00007F"))
# corrplot(raw.cor,
#          type="lower", order="original",
#          col=col1(10),
#          method = 'circle',
#          is.corr = FALSE)

##### SEPARATE DATA
aw.raw <- rawcounts[,str_detect(colnames(rawcounts), 'trans') |
                       str_detect(colnames(rawcounts), 'AW')]
aw.raw[,-1] <- sapply(aw.raw[,-1], as.integer)
rownames(aw.raw) <- aw.raw$transcript
aw.raw <- aw.raw[,-1]

ybr1.raw <- rawcounts[,str_detect(colnames(rawcounts), 'trans') |
                      (str_detect(colnames(rawcounts), 'SP') &
                         str_detect(colnames(rawcounts), 'One'))]
ybr1.raw[,-1] <- sapply(ybr1.raw[,-1], as.integer)
rownames(ybr1.raw) <- ybr1.raw$transcript
ybr1.raw <- ybr1.raw[,-1]

ybr2.raw <- rawcounts[,str_detect(colnames(rawcounts), 'trans') |
                        (str_detect(colnames(rawcounts), 'SP') &
                           str_detect(colnames(rawcounts), 'Two'))]
ybr2.raw[,-1] <- sapply(ybr2.raw[,-1], as.integer)
rownames(ybr2.raw) <- ybr2.raw$transcript
ybr2.raw <- ybr2.raw[,-1]


##### CLEAN DATA
sprintf('There are %d genes with more than 10 reads in total.' ,sum(rowSums(ybr1.raw) >= 10))
sprintf('There are %d genes with more than 10 reads/sample.' ,sum(rowSums(ybr1.raw < 10) == 0))

keep <- rowSums(ybr1.raw) >= 10
ybr1.raw <- ybr1.raw[keep,]

##### NORMALIZING COUNTS
ybr1.mat <- coldata[coldata$owner == 'SP' & coldata$attempt == 'One',]
ybr1.mat$name <- factor(ybr1.mat$name, levels = c('WT','KAN','CRISPY','ATG'))
ybr1.dds <- DESeqDataSetFromMatrix(countData = ybr1.raw,
                              colData = ybr1.mat,
                              design = ~name)
ybr1.dds <- DESeq(ybr1.dds)
ybr1.ncnts <- counts(ybr1.dds, normalized=TRUE)


# ybr2.mat <- coldata[coldata$owner == 'SP' & coldata$attempt == 'Two',]
# ybr2.mat$name <- factor(ybr2.mat$name, levels = c('PAM1','WT','PAM2','CRISPY','ATG','OE'))
# ybr2.dds <- DESeqDataSetFromMatrix(countData = as.matrix(ybr2.raw[,c(2,12,13,
#                                                                      14,15,16,
#                                                                      17,18,19,
#                                                                      6,7,8,
#                                                                      3,4,5,
#                                                                      9,10,11)-1]),
#                                    colData = ybr2.mat,
#                                    design = ~name)
# ybr2.dds <- DESeq(ybr2.dds)
# ybr2.ncnts <- counts(ybr2.dds, normalized=TRUE)

##### PCA ANALYSIS
ybr1.vsd <- vst(ybr1.dds)
plotPCA(ybr1.vsd, intgroup = c("name")) +
  geom_text(aes(label = name), position = position_nudge(y = 2))

# ybr2.vsd <- vst(ybr2.dds)
# plotPCA(ybr2.vsd, intgroup = c("name")) +
#   geom_text(aes(label = name), position = position_nudge(y = 2))


##### DIFFERENTIAL ANALYSIS
ybr1.res <- data.frame()
for (c in resultsNames(ybr1.dds)[2:length(resultsNames(ybr1.dds))]){
  # cat(sprintf('Log fold change, DE, GO and KEGG results for %s.\n',str_remove(c, 'name_')))
  cat(sprintf('Processing %s...\n',str_remove(c, 'name_')))
  temp <- lfcShrink(ybr1.dds, coef=c, type="apeglm")
  DESeq2::plotMA(temp, ylim = c(-5, 5), main=str_remove(c, 'name_'))
  cat(sprintf('When comparing %s there are %d transcripts with more than 1 log fold change at FDR of 5%%. With %d upregulated and %d downregulated.\n\n',
              str_remove(c, 'name_'),
              sum(temp$padj < 0.05 & abs(temp$log2FoldChange) >= 1, na.rm = T),
              sum(temp$padj < 0.05 & temp$log2FoldChange >= 1, na.rm = T),
              sum(temp$padj < 0.05 & temp$log2FoldChange <= -1, na.rm = T)))
  temp$orf_name <- rownames(temp)
  rownames(temp) <- NULL
  ybr1.res <- rbind(ybr1.res, data.frame(temp, contrast = str_remove(c, 'name_')))
}
head(ybr1.res)

ybr1.res <- merge(orfs[,c('gene_systematic_name','transcript','orf_id')], ybr1.res, by.x = 'transcript', by.y = 'orf_name')

ybr1.res %>%
  filter(contrast == 'ATG_vs_WT', abs(log2FoldChange) > 1, gene_systematic_name == 'X')
ybr1.res %>%
  filter(orf_id %in% c(143103, 143104, 143113, 143114))


# ybr2.res <- data.frame()
# for (c in resultsNames(ybr2.dds)[2:length(resultsNames(ybr2.dds))]){
#   # cat(sprintf('Log fold change, DE, GO and KEGG results for %s.\n',str_remove(c, 'name_')))
#   cat(sprintf('Processing %s...\n',str_remove(c, 'name_')))
#   temp <- lfcShrink(ybr2.dds, coef=c, type="apeglm")
#   DESeq2::plotMA(temp, ylim = c(-5, 5), main=str_remove(c, 'name_'))
#   cat(sprintf('When comparing %s there are %d transcripts with more than 1 log fold change at FDR of 5%%. With %d upregulated and %d downregulated.\n\n',
#               str_remove(c, 'name_'),
#               sum(temp$padj < 0.05 & abs(temp$log2FoldChange) >= 1, na.rm = T),
#               sum(temp$padj < 0.05 & temp$log2FoldChange >= 1, na.rm = T),
#               sum(temp$padj < 0.05 & temp$log2FoldChange <= -1, na.rm = T)))
#   temp$orf_name <- rownames(temp)
#   rownames(temp) <- NULL
#   ybr2.res <- rbind(ybr2.res, data.frame(temp, contrast = str_remove(c, 'name_')))
# }
# head(ybr2.res)
# 
# ybr2.res <- merge(orfs[,c('gene_systematic_name','transcript','orf_id')], ybr2.res, by.x = 'transcript', by.y = 'orf_name')
# 
# ybr2.res %>%
#   filter(contrast == 'ATG_vs_PAM1', abs(log2FoldChange) > 1, gene_systematic_name == 'X')
# ybr2.res %>%
#   filter(orf_id %in% c(143103, 143104, 143113, 143114))


##### COMBINING COEX COIP AND RNASEQ DATA

## YBR1
head(ybr1.res)
head(ybr.coex)

ybr1.res <- merge(ybr.coex, ybr1.res,
      by = 'transcript',
      all.y = T)

ybr1.res <- ybr1.res[order(ybr1.res$contrast, ybr1.res$log2FoldChange, ybr1.res$rho, decreasing = T),]

ybr1.res %>%
  filter(contrast == 'ATG_vs_WT')

melt(quantile(ybr.coex$rho, c(0.01,0.05,0.5,0.95,0.99), na.rm = T))

plot.ybr1.coex.ex <- ybr1.res %>%
  filter(padj <= 0.05) %>%
  ggplot(aes(x = rho, y = log2FoldChange)) +
  geom_vline(xintercept = melt(quantile(ybr.coex$rho, c(0.5,0.95,0.99), na.rm = T))[[1]],
             linetype = 'dashed', lwd = 0.2) +
  geom_point(size = 0.5, alpha = 0.5, shape = 21) +
  geom_point(data = ybr1.res %>%
               filter(!is.na(coip_rank), padj <= 0.05),
             aes(x = rho, y = log2FoldChange, col = "CoIP'd with YBR"),
             size = 0.5, shape = 19) +
  # geom_smooth(data = ybr1.res %>%
  #              filter(!is.na(coip_rank), padj <= 0.05),
  #            aes(x = rho, y = log2FoldChange, col = "CoIP'd with YBR"),
  #            method = 'lm') +
  geom_point(data = ybr1.res %>%
               filter(gene == 'YBR196C'),
             aes(x = rho, y = log2FoldChange, col = "PGI1"),
             size = 0.5, shape = 19) +
  geom_point(data = ybr1.res %>%
               filter(orf_id %in% c(143103, 143104, 143113, 143114)),
             aes(x = rho, y = log2FoldChange, col = "YBR Friends"),
             size = 0.5, shape = 19) +
  geom_point(data = ybr1.res %>%
               filter(gene == 'YBR196C-A'),
             aes(x = rho, y = log2FoldChange, col = 'YBR196C-A'),
             size = 0.5, shape = 19) +
  geom_smooth(method = 'lm') +
  stat_cor(method = 'spearman', size = 2) +
  scale_y_continuous(breaks = seq(-10,10,1)) +
  facet_wrap(.~contrast) +
  theme_linedraw() +
  theme(axis.title = element_text(size = titles),
        axis.text = element_text(size = txt),
        legend.title = element_blank(),
        legend.text = element_text(size = txt),
        legend.position = 'bottom',
        legend.key.size = unit(3, "mm"),
        legend.box.spacing = unit(0.5,"mm"),
        strip.text = element_text(size = txt,
                                  face = 'bold',
                                  margin = margin(0.1,0,0.1,0, "mm")))
ggsave(sprintf("%s/YBR_RNASEQ1_COEX.jpg",fig_path), plot.ybr1.coex.ex,
       height = one.5c, width = two.c, units = 'mm',
       bg = 'white',
       dpi = 600)


ybr1.res %>%
  filter(orf_id %in% c(143103, 143104, 143113, 143114))
         
ybr1.ncnts %>%
  data.frame() %>%
  mutate(transcript = rownames(ybr1.ncnts)) %>%
  melt(id.vars = 'transcript') %>%
  mutate(sample = str_split(variable, '_', simplify = T)[,1]) %>%
  filter(transcript %in% c('chr2_614215','chr2_614702','chr2_614936','chr2_614024')) %>%
  group_by(sample, transcript) %>%
  data.frame() %>%
  ggplot(aes(x = sample, y = value)) +
  geom_boxplot() +
  facet_wrap(.~transcript)

ybr1.raw %>%
  data.frame() %>%
  mutate(transcript = rownames(ybr1.raw)) %>%
  melt(id.vars = 'transcript') %>%
  mutate(sample = str_split(variable, '_', simplify = T)[,1]) %>%
  filter(transcript %in% c('chr2_614215','chr2_614702','chr2_614936','chr2_614024')) %>%
  group_by(sample, transcript) %>%
  data.frame() %>%
  ggplot(aes(x = sample, y = value)) +
  geom_boxplot() +
  facet_wrap(.~transcript)

ybr1.res %>%
  filter(!is.na(coip_rank), padj <= 0.05, log2FoldChange < 0, contrast == 'ATG_vs_WT') 


# ybr1.lfc <- ybr.coex[,c('gene','transcript','classification','rho','coip_rank')]
# for (c in unique(ybr1.res$contrast)) {
#   temp <- ybr1.res[ybr1.res$contrast == c, c('transcript','log2FoldChange')]
#   ybr1.lfc <- merge(ybr1.lfc, temp, by = 'transcript', all = T)
#   colnames(ybr1.lfc) <- c(colnames(ybr1.lfc[,-length(ybr1.lfc)]), c)
# }
# 
# ybr1.lfc %>%
#   ggplot(aes(x = ATG_vs_WT, y = CRISPY_vs_WT)) +
#   geom_point() +
#   geom_smooth(method = 'lm')

# ## YBR2
# head(ybr1.res)
# head(ybr.coex)
# 
# ybr2.res <- merge(ybr.coex, ybr2.res,
#                   by.x = c('transcript','gene'),
#                   by.y = c('transcript','gene_systematic_name'),
#                   all.y = T)
# ybr2.res <- ybr2.res[order(ybr2.res$contrast, ybr2.res$log2FoldChange, ybr2.res$rho, decreasing = T),]
# 
# ybr2.res %>%
#   ggplot(aes(x = rho, y = log2FoldChange)) +
#   geom_point(size = 0.5, alpha = 0.5) +
#   geom_density2d() +
#   geom_smooth(method = 'lm') +
#   facet_wrap(.~contrast)
# 
# 
# ybr2.lfc <- ybr.coex[,c('gene','transcript','classification','rho','coip_rank')]
# for (c in unique(ybr2.res$contrast)) {
#   temp <- ybr2.res[ybr2.res$contrast == c, c('transcript','log2FoldChange')]
#   ybr2.lfc <- merge(ybr2.lfc, temp, by = 'transcript', all = T)
#   colnames(ybr2.lfc) <- c(colnames(ybr2.lfc[,-length(ybr2.lfc)]), c)
# }
# 
# ybr2.lfc %>%
#   ggplot(aes(x = ATG_vs_PAM1, y = CRISPY_vs_PAM1)) +
#   geom_point() +
#   geom_smooth(method = 'lm')

##### GO/KEGG ENRICHMENT
allgenes <- ybr1.res$gene[ybr1.res$padj <= 0.05 & ybr1.res$contrast == 'ATG_vs_WT']
allgenes <- bitr(allgenes, fromType = "ORF",
                 toType = c("ENTREZID","GENENAME","ENSEMBL"),
                 OrgDb = org.Sc.sgd.db)

orfoi <- ybr1.res$gene[!is.na(ybr1.res$coip_rank) & ybr1.res$padj <= 0.05 & ybr1.res$log2FoldChange > 0 & ybr1.res$contrast == 'ATG_vs_WT']
orfoi <- bitr(orfoi, fromType = "ORF",
                 toType = c("ENTREZID","GENENAME","ENSEMBL"),
                 OrgDb = org.Sc.sgd.db)
orfoi <- orfoi[!is.na(orfoi$ENSEMBL),]

goe <- enrichGO(gene          = orfoi$ENSEMBL,
                universe      = allgenes$ENSEMBL,
                OrgDb         = org.Sc.sgd.db,
                keyType       = "ENSEMBL",
                ont           = "ALL",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05)

kegg <- enrichKEGG(gene         = orfoi$ENSEMBL,
                   universe     = allgenes$ENSEMBL,
                   organism     = 'sce',
                   pvalueCutoff = 0.05)
data.frame(goe)


##### ERMES, vacuole, etc. related ORFs
vatpase.orfs <- c('YGR106C')
ermes.orfs <- c('YER178W','YLR253W','YAL048C','YLL006W','YOL009C','YFL018C','YAL010C','YOR228C','YGL219C','YAL008W')


ybr1.res %>%
  filter(contrast == 'ATG_vs_WT', gene %in% c('YGL167C','YGL008C','YBR109C','YBR187W','YJR106W','YDR040C'))




